### PR TITLE
UPSTREAM: <carry>: CFE-910: Update route external certificate validation

### PR DESCRIPTION
**Pull request by thejasn**
Description
---
Updates route validations based on https://github.com/openshift/enhancements/pull/1307 which introduces a new field in the route API `externalCertificate` behind the TP feature gate. 

Injects openshift feature gates into the custom validator for routes, similar to https://github.com/openshift/openshift-apiserver/pull/382. Openshift feature gates were already being passed by the operator in https://github.com/openshift/cluster-kube-apiserver-operator/pull/1485, this is a follow up for consuming the feature gates for route validation. 

Follow up from: https://github.com/openshift/library-go/pull/1549

### CODE REVIEW

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 1]
```diff
@@ -45,18 +41,21 @@ func toRoute(uncastObj runtime.Object) (*routev1.Route, field.ErrorList) {
 }
 
 type routeV1 struct {
+	sarc           authorizationv1client.SubjectAccessReviewInterface
+	secretsGetter  corev1client.SecretsGetter
+	validationOpts routecommon.RouteValidationOptions
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:46]: If `routecommon.RouteValidationOptions` is a struct then I think you will need a function.

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 2]
```diff
@@ -45,18 +41,21 @@ func toRoute(uncastObj runtime.Object) (*routev1.Route, field.ErrorList) {
 }
 
 type routeV1 struct {
+	sarc           authorizationv1client.SubjectAccessReviewInterface
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:44]: I think that you will need a function here as well.

Comment (p0lyn0mial) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:44]: same for `secretsGetter`

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+type validateCustomResourceWithClient struct {
+	admission.ValidationInterface
+
+	secretGetter        corev1client.SecretsGetter
+	sarc                authorizationv1client.SubjectAccessReviewInterface
+	routeValidationOpts routecommon.RouteValidationOptions
+}
+
+func NewValidateRoute() (admission.Interface, error) {
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go:29]: could you please add a unit test that would ensue that the `getters` can be injected?

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	a.sarc = authclient.SubjectAccessReviews()
+
+	a.routeValidationOpts = routecommon.RouteValidationOptions{
+		AllowExternalCertificates: feature.DefaultMutableFeatureGate.Enabled(featuregate.Feature(configv1.FeatureGateRouteExternalCertificate)),
+	}
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go:84]: should we check if `a.ValidationInterface` also wants `SetRESTClientConfig`?

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go [hunk 5]
```diff
@@ -45,18 +41,21 @@ func toRoute(uncastObj runtime.Object) (*routev1.Route, field.ErrorList) {
...
 
-func (routeV1) ValidateCreate(obj runtime.Object) field.ErrorList {
+func (r routeV1) ValidateCreate(obj runtime.Object) field.ErrorList {
 	routeObj, errs := toRoute(obj)
 	if len(errs) > 0 {
 		return errs
 	}
 
-	return routevalidation.ValidateRoute(routeObj)
+	return routevalidation.ValidateRoute(context.TODO(), routeObj, r.sarc, r.secretsGetter, r.validationOpts)
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:55]: I think that we should extend `ObjectValidator` to accept the `ctx` for all its methods.

```
type ObjectValidator interface {
	ValidateCreate(obj runtime.Object) field.ErrorList
	ValidateUpdate(obj runtime.Object, oldObj runtime.Object) field.ErrorList
	ValidateStatusUpdate(obj runtime.Object, oldObj runtime.Object) field.ErrorList
}
```

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	authclient, err := authorizationv1client.NewForConfig(&restClientConfig)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	a.sarc = authclient.SubjectAccessReviews()
+
+	a.routeValidationOpts = routecommon.RouteValidationOptions{
+		AllowExternalCertificates: feature.DefaultMutableFeatureGate.Enabled(featuregate.Feature(configv1.FeatureGateRouteExternalCertificate)),
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go:83]: How hard would it be to move this initialisation to a separate setter ?

As we did for https://github.com/openshift/kubernetes/blob/dee15e888b06a2a1c1ac5a4c10f0cba9077e8d57/openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go#L113

File: openshift-kube-apiserver/admission/route/hostassignment/admission.go [hunk 7]
```diff
@@ -147,6 +152,9 @@ var _ initializer.WantsExternalKubeClientSet = &hostAssignment{}
 
 func (a *hostAssignment) SetExternalKubeClientSet(clientset kubernetes.Interface) {
 	a.sarClient = clientset.AuthorizationV1().SubjectAccessReviews()
+	a.validationOpts = routecommon.RouteValidationOptions{
```
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/route/hostassignment/admission.go:155]: Same question here.

How hard would it be to move this initialisation to a separate setter ?

As we did for https://github.com/openshift/kubernetes/blob/dee15e888b06a2a1c1ac5a4c10f0cba9077e8d57/openshift-kube-apiserver/admission/scheduler/nodeenv/admission.go#L113

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 2 above)
Comment (p0lyn0mial) [inline openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:44]: BTW: Perhaps this question should be asked elsewhere. But could you explain why we need the `SAR` client? It seems to me that we want to make sure if the user can change `Route.Spec.Host`, right? Do you happen to know why this was implemented in this particular way?

**Review by p0lyn0mial — COMMENTED** [review]
I added a few comments. 
Please also check why all CI jobs failed.

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 2 above)
Comment (p0lyn0mial) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go:44]: /cc @benluddy

File: openshift-kube-apiserver/openshiftkubeapiserver/flags.go [hunk 8]
```diff
@@ -42,6 +45,15 @@ func ConfigToFlags(kubeAPIServerConfig *kubecontrolplanev1.KubeAPIServerConfig)
 	configflags.SetIfUnset(args, "tls-min-version", kubeAPIServerConfig.ServingInfo.MinTLSVersion)
 	configflags.SetIfUnset(args, "tls-sni-cert-key", sniCertKeys(kubeAPIServerConfig.ServingInfo.NamedCertificates)...)
 
+	// set the global featuregates.`
+	warnings, err := features.SetFeatureGates(args, feature.DefaultMutableFeatureGate)
```
[the remark this answers is not in the document]
Comment (thejasn) [reply openshift-kube-apiserver/openshiftkubeapiserver/flags.go:49]: k8s apiserver never had the OCP feature flags being injected

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 1 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go]: Addressed

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 2 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go]: Addressed

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 5 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go]: Addressed

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go (hunk 6 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validation_wrapper.go]: Addressed

File: openshift-kube-apiserver/admission/route/hostassignment/admission.go (hunk 7 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/route/hostassignment/admission.go]: Addressed

File: openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go (hunk 2 above)
Comment (thejasn) [reply openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go]: The SAR client is specifically needed by the host update validation (https://github.com/openshift/library-go/blob/c91dd9756953b9b7ff3139565886954ed0cc33d3/pkg/route/hostassignment/assignment.go#L120). It is used to verify rbac on the `custom-host` sub resource. I am re-using the SAR client to also validate permissions needed for reading secrets (https://github.com/openshift/enhancements/blob/master/enhancements/ingress/route-secret-injection-for-external-certificate-management.md#implementation-detailsnotesconstraints-optional).

**Comment by benluddy** [comment]
Are E2E tests covering the validation changes passing for MicroShift? Since none of the other flavors use the kube-apiserver admission plugin to validate routes, the presubmits give limited signal.

**Pull request closed without merging**